server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java

```diff
@@ -791,6 +798,50 @@ public void run() {
       }
     }, 0, 5000, TimeUnit.MILLISECONDS);
 
+    // Periodically check that metadata of tablets matches what is held in memory
+    ThreadPools.createGeneralScheduledExecutorService(aconf).scheduleWithFixedDelay(() -> {
+      final SortedMap<KeyExtent,Tablet> onlineTabletsSnapshot = onlineTablets.snapshot();
+
+      final SortedSet<KeyExtent> userTablets = new TreeSet<>();
+      final SortedSet<KeyExtent> nonUserTablets = new TreeSet<>();
+
+      // Create subsets of tablets based on DataLevel: one set who's DataLevel is USER and another
+      // containing the remaining tablets (those who's DataLevel is ROOT or METADATA).
+      // This needs to happen so we can use .readTablets() on the DataLevel.USER tablets in order
+      // to reduce RPCs.
+      // TODO: Push this partitioning, based on DataLevel, to ample.
+      onlineTabletsSnapshot.forEach((k, v) -> {
+        if (Ample.DataLevel.of(k.tableId()) == Ample.DataLevel.USER) {
+          userTablets.add(k);
+        } else {
+          nonUserTablets.add(k);
+        }
+      });
+
+      List<TabletMetadata> tmdList;
+
+      // gather metadata for all tablets with DataLevel.USER using readTablets()
+      try (TabletsMetadata tabletsMetadata = getContext().getAmple().readTablets()
```

Comment:
I think ample used to read tablets using only the end row, so could get back things w/ diff prev end row than in key extent.  However I think we changed Ample to only return data where the end row and prev end row match.  But not completely sure, need to verify that as the correctness of this code depends on it.

Reply:
If ample should only return key extents that we ask for, then seeing anything outside of that set would indicate an error in ample.  Not sure if we want to bother checking for that error case here, but we could possibly do that.

Reply:
Yea, in #2189 it was made so that "the end row and prev end row exactly match the given extent".